### PR TITLE
Add audmath.window()

### PR DESCRIPTION
**Pull request by hagenw**
Adds `audmath.window()` to create a (half-)window that can be used as a fade-in.

image

### CODE REVIEW

**Comment by frankenjoe** [comment]
General question: wouldn't it make sense to start with a `window()` function that generates a complete window and then add an argument to retrieve the full window, or just the left / right part of it?

**Comment by hagenw** [comment]
My main reason to not start with a window function was that most of them exists already, e.g. `np.kaiser`, `scipy.signal.windows.tukey`. So I thought it would be more handy to provide directly a fade-in here.

But I think you are right, and it would be more general to add `audmath.window()`. I chnaged it accordingly, and added a `half=None` argument, which can be used to select a specific half-window.

**Comment by hagenw** [comment]
As we now have a window and not a fade-in, we might also consider providing `level` not in decibel, but directly as a value and remove `bottom`?

**Comment by frankenjoe** [comment]
> As we now have a window and not a fade-in, we might also consider providing `level` not in decibel, but directly as a value and remove `bottom`?

Yes, makes sense.

**Comment by hagenw** [comment]
> > As we now have a window and not a fade-in, we might also consider providing `level` not in decibel, but directly as a value and remove `bottom`?
> 
> Yes, makes sense.

I replaced `level` and `bottom` by `offset`.

File: audmath/core/api.py [hunk 1]
```diff
@@ -78,6 +87,152 @@ def db(
     return x
 
 
+def window(
+    samples: int,
+    shape: str = 'tukey',
+    half: str = None,
+    offset: typing.Union[int, float] = 0,
```
Comment (frankenjoe) [inline audmath/core/api.py:94]: Since we have an `offset` would it also make sense to have an option to set the peak?

Comment (hagenw) [reply audmath/core/api.py:94]: Mh, good question.
I think I would rather remove the `offset` argument and leaf the scaling to the user as it is trivial anyway, e.g.
```
win * (1. - offset) + offset
```

Comment (frankenjoe) [reply audmath/core/api.py:94]: Agree, let's keep it simple.

Comment (hagenw) [reply audmath/core/api.py]: I removed the `offset` argument.

File: audmath/core/api.py [hunk 2]
```diff
@@ -78,6 +87,141 @@ def db(
...
+        fig.set_size_inches(6.4, 3.84)
+        plt.tight_layout()
+
+    Args:
+        samples: length of window
+        shape: shape of window
+        half: if ``None`` return whole window,
+            if ``left`` or ``right``
+            return left or right window
+            inlcuding center
```
Comment (hagenw) [inline audmath/core/api.py:144]: I guess I need to reformulate this. At the moment if you specify `left` or `right` the half-window will always include 1 as maximum, whereas if you specify `half=None` and request a window with an odd number of samples you might get a window without 1 as maximum. This means in the case of an odd number of samples we also have `np.concatenate([window(samples / 2, half=left) + window(samples / 2, half=right]) != window(samples)`

Comment (hagenw) [reply audmath/core/api.py:144]: I decided for this as it is very handy to always get a maximum of 1 for a half window. But if you disagree and it is more important that the two half-windows would combine to the whole window, we can also change this.

Comment (frankenjoe) [reply audmath/core/api.py:144]: With the use case we have in mind (applying a half window as fade in / out) it makes sense to include 1 for half windows. So I think we can stick with your solution, but should extend the docstring to be clear about it.

Comment (hagenw) [reply audmath/core/api.py]: I updated the text to

image

**Pull request merged**